content/named-tuples.md

```diff
@@ -0,0 +1,777 @@
+`apply`, `head`, `tail`, `take`, `drop`, `++`, `map`, or `zip`.
+Similar to `Tuple`, the `NamedTuple` object also contains types such as `Elem`, `Head`, `Concat`
+that describe the results of these extension methods.
+
+The translation of named tuples to instances of `NamedTuple` is fixed by the specification and therefore known to the programmer. This means that:
+
+ - All tuple operations also work with named tuples "out of the box".
+ - Macro libraries can rely on this expansion.
+
+### Computed Field Names
+
+The `Selectable` trait now has a `Fields` type member that can be instantiated
+to a named tuple.
+
+```scala
+trait Selectable:
+  type Fields <: NamedTuple.AnyNamedTuple
+```
+
+If `Fields` is instantiated in a subclass of `Selectable` to some named tuple type,
+then the available fields and their types will be defined by that type. Assume `n: T`
+is an element of the `Fields` type in some class `C` that implements `Selectable`,
+that `c: C`, and that `n` is not otherwise legal as a name of a selection on `c`.
+Then `c.n` is a legal selection, which expands to `c.selectDynamic("n").asInstanceOf[T]`.
+
+It is the task of the implementation of `selectDynamic` in `C` to ensure that its
+computed result conforms to the predicted type `T`
+
+As an example, assume we have a query type `Q[T]` defined as follows:
+
+```scala
+trait Q[T] extends Selectable:
+  type Fields = NamedTuple.Map[NamedTuple.From[T], Q]
+  def selectDynamic(fieldName: String) = ...
+```
+
+Assume in the user domain:
+```scala
+case class City(zipCode: Int, name: String, population: Int)
+val city: Q[City]
+```
+Then
+```scala
+city.zipCode
+```
+has type `Q[Int]` and it expands to
+```scala
+city.selectDynamic("zipCode").asInstanceOf[Q[Int]]
+```
+
+### The NamedTuple.From Type
+
+The `NamedTuple` object contains a type definition
+```scala
+  type From[T] <: AnyNamedTuple
+```
+`From` is treated specially by the compiler. When `NamedTuple.From` is applied to
+an argument type that is an instance of a case class, the type expands to the named
+tuple consisting of all the fields of that case class.
+Here, _fields_ means: elements of the first parameter section. For instance, assuming
+```scala
+case class City(zip: Int, name: String, population: Int)
+```
+then `NamedTuple.From[City]` is the named tuple
+```scala
+(zip: Int, name: String, population: Int)
+```
+The same works for enum cases expanding to case classes, abstract types with case classes as upper bound, alias types expanding to case classes
+and singleton types with case classes as underlying type (in terms of the implementation, the `classSymbol` of a type must be a case class).
+
+`From` is also defined on named tuples. If `NT` is a named tuple type, then `From[NT] = NT`.
+
+### Pattern Matching with Named Fields in General
+
+We allow named patterns not just for named tuples but also for case classes. For instance:
```

Comment:
How does this work? Wouldn't we have to replace 

```scala
def unapply(x: CaseClass): CaseClass
```

with 

```scala
def unapply(x: CaseClass): (foo: Bar, qux: Baz, ...)
```

? 

Or is this more generic, and works with anything returned from `def unapply` as long as the type has a zero-param-list member defined with the correct name?

Reply:
> Or is this more generic, and works with anything returned from def unapply as long as the type has a zero-param-list member defined with the correct name?

Correct.

Reply:
What about members that take implicits? I'm imagining e.g. the Dotty internal types, where just about everything takes a `(given Context)`, even though for most intents and purposes I think of them as fields

What if the return type of `unapply` is a Java type? Do we accept named pattern matching on zero-parameter "getter" methods, or only on Java fields? Or do we prohibit named pattern matching on Java types entirely?

Reply:
Implicits are handled by the existing rules. To support a named pattern match, an `unapply` must return one of the following (after handling the get/isDefined cases)

 - A reference to a case class. The fields are the fields in the first parameter list of the case class. (Note that implicits come later)
 - A reference to a named tuple.

Reply:
What is the reason we want to limit the return type of `unapply` to case-classes and named-tuples types here? It seems the "look up properties on type returned by `unapply` would work for normal `class`es or `trait` values as well?

As a point of contrast Python's structural pattern matching on class instances works across arbitrary types ([link](https://docs.python.org/3/reference/compound_stmts.html#class-patterns)), as long as they have the relevant properties that are being matched against.

Reply:
> Allow unapply to return values of any type, not just case classes or named tuples, and then proceed unchanged with looking up the matched fields on the returned value

That's in a sense already true. `unapply` allows to return instances of any `Product` type. It then proceeds to select the fields of that product type by the `_1`, `_2` selectors. We have a link between these selectors and actual field names only for case classes, since the `_1` selectors are auto-generated for them. For other `Product` instances, these things can be user defined in arbitrary ways so we don't have that info.

Another, deeper reason why pattern matching does not work out of the box for normal classes is that in Scala the constructor elements of a normal class are retained only if prefixed by `val`. So there might be nothing to select. E.g.

```
class Point(x: Int, y: Int)

p match
  case Point(x, y) // does not work, `x` and `y` are not fields of the class.
```
We could consider relaxing that restriction, so that we also allow pattern matching on all classes that have only `val` parameters in their first section. But I believe that's out of the scope for this SIP since it has nothing to do with named matching.

Reply:
The reason I think it's relevant to named matching is because the idea was to allow something like this

```scala
class Point(x: Int, y: Int){
  val z = x + y
}

p match{
  case Point(z = 0) => 
}
```

This doesn't work with positional pattern matching, because that relies on the `_1` `_2` `_3` selectors, which do not exist for most non-case classes. Named pattern matching relaxes the reliance: we can just go directly to pattern matching on the member named `.z` directly. Previously such a thing didn't make sense, but with named pattern matching it does

Reply:
The main issue with enabling non-product (fields) pattern matching is the complete loss of exhaustivity indicators.

Reply:
@haoyi I see. That use case makes sense. Today we would write
```scala
p match
  case p: Point if p.z == 0 =>
```
The named pattern notation gets more convenient if there are more named elements like this. 

Still, it's a rather big change to the normal pattern matching rules, so I'd prefer we leave that for a possible future time.